Update docs around DNSSEC (#345)
canonical/ubuntu-server-documentation: explanation/dnssec/dnssec.md

For a more detailed explanation of how the DNSSEC validation is performed, please refer to the [Simplified 12-step DNSSEC validation process](https://bind9.readthedocs.io/en/latest/dnssec-guide.html#the-12-step-dnssec-validation-process-simplified) guide from ISC.

## DNS daemons

> Ubuntu supports multiple DNS resolvers, covering a variety of usecases. Most of them support DNSSEC validation, but it might not be activated and set-up with valid trust-anchors automatically.

Reviewer: ```suggestion
Ubuntu supports multiple DNS resolvers, covering a variety of use cases. Most of them support DNSSEC validation, but it might not be activated and set up with valid trust-anchors automatically.
```

<!-- Using non-breaking hyphen & non-breaking space for improved table spacing. -->
| Daemon | Type | DNSSEC support |
| --- | --- | --- |
| systemd&#8209;resolved | Stub&nbsp;Resolver&nbsp;(local) | Yes. Enabled by default via `DNSSEC=allow-downgrade` setting. |
| dnsmasq | Stub Resolver | Yes. Disabled by default. Controlled via `dnssec` and `conf-file=../trust-anchors.conf` settings. |
| bind9 | Recursive Resolver | Yes. Enabled by default via `dnssec-validation auto;` setting. |
<!-- TODO: What about "unbound"? -->

> The **systemd-resolved** stub resolver is pre-installed by default and comes with DNSSEC enabled in fallback mode (as of Ubuntu 25.10). This mode configures the `DNSSEC=allow-downgrade` setting in `/usr/lib/systemd/resolved.conf.d/00-enable-dnssec.conf` (installed by the *systemd-resolved-dnssec* package) and tries to validate the DNSSEC records whenever possible, but at the same time accepts unsigned responses for backwards compatibility with unsigned zones.

Reviewer: ```suggestion
The **systemd-resolved** stub resolver is pre-installed by default and comes with DNSSEC enabled in fallback mode (as of Ubuntu 25.10). This mode configures the `DNSSEC=allow-downgrade` setting in `/usr/lib/systemd/resolved.conf.d/00-enable-dnssec.conf` (installed by the `systemd-resolved-dnssec` package) and tries to validate the DNSSEC records whenever possible, but at the same time accepts unsigned responses for backwards compatibility with unsigned zones.
```

The functionality of DNSSEC in **systemd-resolved** can be confirmed, using the `dig` command on the local stub resolver at `127.0.0.53:53`, by checking for the existence of the **ad** (Authenticated Data) flag:
```
$ dig @127.0.0.53 isc.org +dnssec
[...]
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 1
```

> Should authenticity of DNS records be a concern to you, it's advised to override the default DNSSEC validation settings through an additional drop-in configuration in `/etc/systemd/resolved.conf.d/10-dnssec.conf`, which would reject any unsigned records, after reloading the configuration:

Reviewer 2: I agree that this needs clarification. The manpage uses the expression "support DNSSEC properly", and that leaves a lot of room for interpretation, both in the context of documentation, but also in what the correct response of the service is. For the "yes" value, it says "If the DNS server does not properly support DNSSEC all validations will fail.".
Similarly, the "allow-downgrade" option also uses the "support DNSSEC properly" expression.
I guess the behavior we all want is: if records are signed on the server, we want that signature to be valid, and reject if not. If records are not signed (or not supposed to be signed), then we won't require a signature.
And finally, there might be systems out there that definitely want to only deal with signed records, and reject any DNS response that is not authenticated. Is that what "yes" means here?

```
$ cat /etc/systemd/resolved.conf.d/10-dnssec.conf
[Resolve]
DNSSEC=yes

> $ systemctl reload systemd-resolved.service

Reviewer 2: This needs `sudo`.

```
```{warning}
Be aware that enforcing DNSSEC can lead to errors like `DNS_PROBE_FINISHED_NXDOMAIN` in your browser, especially for local, unsigned domains and you would only be able to reach such services by accessing them through their IP address directly.
<!-- TODO: What about DNS64? (https://blog.apnic.net/2016/06/09/lets-talk-ipv6-dns64-dnssec/) -->
```

Once DNSSEC validation is enforced, you should also be able to confirm it through higher level checks in your browser, e.g. using those 3rd party services:
 * https://internet.nl/test-connection/
 * https://wander.science/projects/dns/dnssec-resolver-test/

```{note}

> In case of issues with Domain Name resolution, make sure to remove any drop-in configs for **systemd-resolved**, execute `systemctl reload systemd-resolved.service` and `apt remove systemd-resolved-dnssec`. This will reset any DNSSEC configuration to `DNSSEC=no` and can be confirmed by executing `resolvectl dnssec`.

Author: In fact, those are _only_ drop-ins for resolved.conf.

And I need to re-phrase this whole section, as the `systemd-resolved-dnssec` package is now gone again.

<!-- Using non-breaking hyphen & non-breaking space for improved table spacing. -->
| Daemon | Type | DNSSEC support |
| --- | --- | --- |
| systemd&#8209;resolved | Stub&nbsp;Resolver&nbsp;(local) | Yes. Disabled by default. Controlled via `DNSSEC=...` setting in `/etc/systemd/resolved.conf.d`. |
| dnsmasq | Stub Resolver | Yes. Disabled by default. Controlled via `dnssec` and `conf-file=../trust-anchors.conf` settings. |
| bind9 | Recursive Resolver | Yes. Enabled by default via `dnssec-validation auto;` setting. |
<!-- TODO: What about "unbound"? -->

The **systemd-resolved** stub resolver is pre-installed in Ubuntu but ships with DNSSEC validation disabled by default. It supports two optional DNSSEC modes: fallback or strict validation. Either mode can be configured using the `DNSSEC=` setting. To do this you need to create a directory for custom `resolved.conf.d` drop-in configuration files, set the corresponding setting inside the `[Resolve]` section and reload the configuration, e.g. in `/etc/systemd/resolved.conf.d/10-dnssec.conf`.

> 

Reviewer 2: Given the example and more details below, I think you can drop the very long sentence above which starts with "To do this you need to create a directory ...". But at the same time I can see how someone could become a bit disoriented when told to adjust DNSSEC=, without any indication where, even though that is still coming. I'll leave it up to you. I just found that sentence very long, with several steps. Perhaps a rephrase would also work, and that just said "create this directory" and "this file inside it".

* You can enable the **fallback mode**, using the `DNSSEC=allow-downgrade` setting, which tries to validate the DNSSEC records whenever possible, but at the same time accepts unsigned responses for backwards compatibility with unsigned zones.

> * Should authenticity of DNS records be a concern to you, it's advised to opt-in to the **strict mode** by using the `DNSSEC=yes` setting and reloading the configuration:

Reviewer 2: I would drop the `:` from the sentence above, and start a new paragraph. It would start with something like:
"""
For example, here is how the strict mode can be enabled. To use the fallback mode instead, replace `DNSSEC=yes` with `DNSSEC=allow-downgrade`:
"""
And then we have the example.

Author: +1 – I was pondering about that colon anyway, as it felt odd having it only in one bullet-point.

```
$ sudo mkdir -p /etc/systemd/resolved.conf.d
$ sudo tee /etc/systemd/resolved.conf.d/10-dnssec.conf >/dev/null <<EOF
[Resolve]
DNSSEC=yes
EOF

$ sudo systemctl reload systemd-resolved.service
```

> Once reloaded, the functionality of DNSSEC in **systemd-resolved** can be confirmed, using the `dig` command on the local stub resolver at `127.0.0.53:53`, by checking for the existence of the **ad** (Authenticated Data) flag:

Reviewer 2: Maybe drop the port, as it's not used in the dig command down below. Just a thought, I'm fine to keep it if you prefer.

           DNS Servers: 10.10.17.1
            DNS Domain: lxd

```{note}

> Starting with Ubuntu 25.10 the *systemd-resolved-dnssec* package is pre-installed enabling the fallback mode as `DNSSEC=allow-downgrade/supported`

Reviewer: ```suggestion
Starting with Ubuntu 25.10 the `systemd-resolved-dnssec` package is pre-installed enabling the fallback mode as `DNSSEC=allow-downgrade/supported`
```

### Local DNSSEC validation
Some stub resolvers, such as systemd-resolved, can perform DNSSEC validation locally. This eliminates the risk of network attacks between the resolver and the client, as they reside on the same system. However, local DNSSEC validation introduces additional overhead in the form of multiple DNS queries. For each DNS query, the resolver must fetch the desired record, its digital signature, and the corresponding public key. This process can increase latency, and with multiple clients on the same network request the same record, that's duplicated work.

In general, local DNSSEC validation is still the more secure approach, validating and authenticating the DNS resource records end-to-end, without the need to trust any DNS server along the way. Besides this, DNS-over-TLS (DoT) or DNS-over-HTTPS (DoH) could be used to increase privacy, by encrypting the DNS connection between your local client and the remote Recursive Resolver.

As an example, let's perform the same query using `systemd-resolved` with and without local DNSSEC validation enabled.

Without local DNSSEC validation. First, let's show it's disabled indeed:

>     $ sudo resolvectl dnssec eth0 false

Author: ACK. Adding a new section about resolvectl troubleshooting in the howto area.

In any case, having a Validating Resolver in the network is a valid and very useful scenario, and good enough for most cases. And it has the extra benefit that the DNSSEC validation is done only once, at the resolver, for all clients on the network.

### Local DNSSEC validation
Some stub resolvers, such as systemd-resolved, can perform DNSSEC validation locally. This eliminates the risk of network attacks between the resolver and the client, as they reside on the same system. However, local DNSSEC validation introduces additional overhead in the form of multiple DNS queries. For each DNS query, the resolver must fetch the desired record, its digital signature, and the corresponding public key. This process can increase latency, and with multiple clients on the same network request the same record, that's duplicated work.

> In general, local DNSSEC validation is still the most secure approach, validating the records end-to-end, without the need to trust any DNS server along the way.

Reviewer 3: Since `systemd-resolved` support DoT, it might be worth adding a quick note that this could also be used to secure the resolver <-> client path if the resolver is trusted.

Author: This is a generic note, so we could even mention DoH, too. As both DoT and DoH could be used to increase privacy.
